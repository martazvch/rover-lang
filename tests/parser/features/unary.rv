-1
--1
-5*3
not true
-{add()}
not if true do "yes" else "no"
not -add()
