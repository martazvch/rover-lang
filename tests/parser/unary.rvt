// Tests unary expressions

code:
-1

expect:
[Unary -]
    [Int literal 1]

================

code:
--1

expect:
[Unary -]
    [Unary -]
        [Int literal 1]

================

code:
-5*3

expect:
[Binop *]
    [Unary -]
        [Int literal 5]
    [Int literal 3]
 
================

code:
not true

expect:
[Unary not]
    [Bool literal true]
 
================

code:
-{add()}

expect:
[Unary -]
    [Block]
        [Fn call
            callee:
            [Identifier add]
            args:
        ]
 
================

code:
not if true do "yes" else "no"

expect:
[Unary not]
    [If
        condition:
        [Bool literal true]
        then body:
        [String literal "yes"]
        else body:
        [String literal "no"]
    ]
 
================

code:
not -add()

expect:
[Unary not]
    [Unary -]
        [Fn call
            callee:
            [Identifier add]
            args:
        ]

================

