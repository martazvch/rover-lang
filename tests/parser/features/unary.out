{
    "unary": {
        "op": "-",
        "expr": {
            "literal": "1"
        }
    },
    "unary": {
        "op": "-",
        "expr": {
            "unary": {
                "op": "-",
                "expr": {
                    "literal": "1"
                }
            }
        }
    },
    "binop": {
        "lhs": {
            "unary": {
                "op": "-",
                "expr": {
                    "literal": "5"
                }
            }
        },
        "rhs": {
            "literal": "3"
        },
        "op": "*"
    },
    "unary": {
        "op": "not",
        "expr": {
            "literal": "true"
        }
    },
    "unary": {
        "op": "-",
        "expr": {
            "block": [
                "fn_call": {
                    "callee": {
                        "literal": "add"
                    },
                    "args": []
                }
            ]
        }
    },
    "unary": {
        "op": "not",
        "expr": {
            "if": {
                "condition": {
                    "literal": "true"
                },
                "then": {
                    "literal": "yes"
                },
                "else": {
                    "literal": "no"
                }
            }
        }
    },
    "unary": {
        "op": "not",
        "expr": {
            "unary": {
                "op": "-",
                "expr": {
                    "fn_call": {
                        "callee": {
                            "literal": "add"
                        },
                        "args": []
                    }
                }
            }
        }
    }
}
